### PR TITLE
Enable tensor test

### PR DESCRIPTION
**Pull request by Canpio**
`paddle::memory` is merged, so I enable tenor's tests.
Because of some interface changes, tensor's current code might be a little rough. I will keep polishing it.

### CODE REVIEW

File: paddle/framework/tensor.h [hunk 1]
```diff
@@ -119,38 +133,37 @@ class Tensor {
...
     virtual void* ptr() const { return static_cast<void*>(ptr_.get()); }
     virtual size_t size() const { return size_; }
-    virtual paddle::platform::Place place() const { return place_; }
+    virtual platform::Place place() const { return place_; }
 
-    std::unique_ptr<T, Deleter> ptr_;
-    paddle::platform::Place place_;  // record the place of ptr_.
-    size_t size_;                    // size of the memory block.
+    std::unique_ptr<T, Deleter<PlaceType>> ptr_;
+    platform::Place place_;  // record the place of ptr_.
```
Comment (QiJune) [inline paddle/framework/tensor.h]: Though ```PlaceholderImpl``` has a template parameter PlaceType, why we need platform::Place here?

Comment (Canpio) [inline paddle/framework/tensor.h]: GPUPlace(0), GPUPlace(1), they are different.

Comment (QiJune) [inline paddle/framework/tensor.h]: GPUPlace and CPUPlace are two different class types. But GPUPlace(0) and GPUPlace(1) are the same class type.

Comment (Canpio) [inline paddle/framework/tensor.h]: Correct. So if we remove `platform::Place place_` and only retain template parameter `PlaceType`, `PlaceholderImpl` is not able to distinguish GPUPlace(0) and GPUPlace(1).

Comment (Canpio) [inline paddle/framework/tensor.h]: I misunderstood Qijun.. Here is a typo, which should be `PlaceType place_`.

Comment (Canpio) [inline paddle/framework/tensor.h]: With `platform::Place` here we can easily use `==` to check two `place` in `mutable_data()` and `data()`.

File: paddle/framework/CMakeLists.txt [hunk 2]
```diff
@@ -2,7 +2,7 @@
 cc_library(ddim SRCS ddim.cc)
 cc_test(ddim_test SRCS ddim_test.cc DEPS ddim)
 nv_test(dim_test SRCS dim_test.cu DEPS ddim)
-cc_test(tensor_test SRCS tensor_test.cc DEPS ddim)
+cc_test(tensor_test SRCS tensor_test.cc DEPS ddim place paddle_memory)
```
Comment (reyoung) [inline paddle/framework/CMakeLists.txt]: Please update develop, because this lines are changed.

File: paddle/framework/tensor.h [hunk 3]
```diff
@@ -119,38 +133,37 @@ class Tensor {
...
-    virtual paddle::platform::Place place() const = 0;
+    virtual platform::Place place() const = 0;
     virtual size_t size() const = 0;
   };
 
-  template <typename T>
+  template <typename T, typename PlaceType>
   struct PlaceholderImpl : public Placeholder {
    private:
+    template <typename PType>
```
Comment (reyoung) [inline paddle/framework/tensor.h]: Is that `PType` useful? Maybe in Deleter, we can also use `PlaceType`

File: paddle/framework/tensor.h [hunk 4]
```diff
@@ -166,7 +179,7 @@ class Tensor {
   DDim dims_;
   size_t numel_;   // cache of `product(dims_)`
   size_t offset_;  // marks the begin of tensor data area.
-};
+};                 // namespace framework
```
Comment (reyoung) [inline paddle/framework/tensor.h]: above line seems to be a typo.

File: paddle/framework/tensor.h [hunk 5]
```diff
@@ -39,21 +37,37 @@ class Tensor {
...
+  T* mutable_data(platform::Place place) {
     PADDLE_ENFORCE(numel_ > 0,
                    "Tensor::numel_ must be larger than zero to call "
                    "Tensor::mutable_data. Call Tensor::set_dim first.");
     if (holder_ == nullptr ||
         !(holder_->place() ==
           place) /* some versions of boost::variant don't have operator!= */
         || holder_->size() < numel_ * sizeof(T) + offset_) {
-      holder_.reset(new PlaceholderImpl<T>(place, numel_ * sizeof(T)));
+#ifdef __CUDACC__
```
Comment (reyoung) [inline paddle/framework/tensor.h]: `#ifdef PADDLE_ONLY_GPU`, use that macro.

File: paddle/framework/tensor.h [hunk 6]
```diff
@@ -39,21 +37,37 @@ class Tensor {
...
+          holder_.reset(new PlaceholderImpl<T, platform::GPUPlace>(
+              boost::get<platform::GPUPlace>(place), numel_ * sizeof(T)));
+          break;
+
+        case 1:
+          holder_.reset(new PlaceholderImpl<T, platform::CPUPlace>(
+              boost::get<platform::CPUPlace>(place), numel_ * sizeof(T)));
+          break;
+      }
+#else
```
Comment (reyoung) [inline paddle/framework/tensor.h]:
```cpp
if (is_cpu_place(place)) {
  holder_.reset(...)
} else if (is_gpu_place(place)) {
#ifdef PADDLE_ONLY_CPU
  PADDLE_THROW("GPU is not enabled");
#else
  holder_.reset(...)
#endif
} else {
  PADDLE_THROW("not supported device.")
}
```

**Pull request merged**